src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -199,14 +210,54 @@ int getTimeout() {
                 MqttClient.DEFAULT_MQTT_OPERATION_TIMEOUT, MqttClient.MQTT_OPERATION_TIMEOUT_KEY));
     }
 
-    private void resubscribe() {
-        subscriptionTopics.forEach((key, value) -> {
-            subscribe(key, value).whenComplete((i, t) -> {
-                if (t != null) {
-                    logger.atError().kv(TOPIC_KEY, key).kv(QOS_KEY, value.name()).log("Unable to resubscribe to topic");
+    private void resubscribe(boolean sessionPresent) {
+        // Don't bother spin up a thread if subscription is empty
+        if (!subscriptionTopics.isEmpty()) {
+            // All subscriptions will be dropped if connected without a session
+            if (!sessionPresent) {
+                droppedSubscriptionTopics.putAll(subscriptionTopics);
+            }
+            if (resubscribeFuture == null || resubscribeFuture.isDone()) {
+                resubscribeFuture = executorService.submit(this::resubscribeDroppedTopicsTask);
+            }
+        }
+    }
+
+    private void resubscribeDroppedTopicsTask() {
+        logger.atInfo().log("Re-subscribing to topics");
+        while (true) {
+            logger.atInfo().kv("failed topics", droppedSubscriptionTopics.keySet())
+                    .kv("sub topics", subscriptionTopics.keySet()).log();
+            Iterator<Map.Entry<String, QualityOfService>> iterator = droppedSubscriptionTopics.entrySet().iterator();
+            while (iterator.hasNext()) {
+                if (!currentlyConnected.get()) {
+                    logger.atInfo().log("Canceling re-subscribe because connection was interrupted");
+                    return;
                 }
-            });
-        });
+
+                Map.Entry<String, QualityOfService> subscriptionEntry = iterator.next();
+                subscribe(subscriptionEntry.getKey(), subscriptionEntry.getValue()).whenComplete((i, t) -> {
+                    if (t == null) {
+                        iterator.remove();
+                    } else {
+                        logger.atError().kv(TOPIC_KEY, subscriptionEntry.getKey())
+                                .kv(QOS_KEY, subscriptionEntry.getValue().name())
+                                .log("Failed to re-subscribe to topic, will retry later");
+                    }
+                });
+            }
+            // If there's still failed subscriptions, wait for some time before trying again
+            if (droppedSubscriptionTopics.isEmpty()) {
+                break;
+            } else {
```

Comment:
This isn't correct since you didn't wait for the futures to complete. The subscription requests will all be sent out, but it is unlikely that you'd have the responses yet.

Reply:
Additionally, since you don't wait for the future to complete, you're going to loop around and then it will send out requests again, even though the first requests could still be outstanding and not actually failed.